Maple2.Server.Game/PacketHandlers/MasteryHandler.cs

```diff
@@ -32,12 +32,52 @@ private enum Command : byte {
     public override void Handle(GameSession session, IByteReader packet) {
         var command = packet.Read<Command>();
         switch (command) {
+            case Command.Reward:
+                HandleReward(session, packet);
+                break;
             case Command.Craft:
                 HandleCraft(session, packet);
                 break;
         }
     }
 
+    private void HandleReward(GameSession session, IByteReader packet) {
+        int rewardBoxDetails = packet.ReadInt();
+        var type = (MasteryType) (rewardBoxDetails / 1000);
+        int level = rewardBoxDetails % 100;
+
+        if (session.Player.Value.Unlock.MasteryRewardsClaimed.TryGetValue(rewardBoxDetails, out bool isClaimed) && isClaimed) {
+            session.Send(MasteryPacket.Error(MasteryError.s_mastery_error_unknown));
+            return;
+        }
+
+        if (!TableMetadata.MasteryRewardTable.Entries.TryGetValue(type, out MasteryRewardTable.Entry? entry) || !entry.Levels.TryGetValue(level, out MasteryRewardTable.Level? levelMetadata)) {
+            return;
+        }
+
+        if (session.Mastery[type] < levelMetadata.Value) {
+            session.Send(MasteryPacket.Error(MasteryError.s_mastery_error_invalid_level));
+            return;
+        }
+
+        if (!ItemMetadata.TryGet(levelMetadata.ItemId, out ItemMetadata? itemMetadata)) {
+            return;
+        }
+
+        var rewardItem = new Item(itemMetadata, levelMetadata.ItemRarity, levelMetadata.ItemAmount);
+
+        if (!session.Item.Inventory.CanAdd(rewardItem)) {
```

Comment:
Rather than checking `CanAdd`, you might as well just call `Add` which returns a bool. You would really only need to use `CanAdd` if there were other things you wanted to check after it, but before actually adding the item.
```cs
if (!session.Item.Inventory.Add(...)) {
    session.Send(ChatPacket.Alert(StringCode.s_err_inventory));
    return;
}

session.Player.Value.Unlock.MasteryRewardsClaimed.Add(rewardBoxDetails, true);
...
```